Fb smc multigrid

ukmo-jianguo-li opened the pull request:
# Pull Request Summary
Add equal ranked SMC multi-grid option.  

## Description
Main changes include: 
1) introduce SMCTYPE grid type so multi-grid can recognise SMC grid type; 
2) merge ARC switch into SMC switch and use logical variable to control Arctic part; 
3) add extra subroutine for equal ranked SMC sub-grids set up in wmgridmd.ftn; 
4) add extra subroutines in w3psmcmd.ftn for boundary point matching among sub-grids; 
5) a few variables for SMC grid are renamed for compatibility with regular grid in multi-grid mode.
6) other related changes for adding SMCTYPE and removing ARC switch.

One bug fixing in wminitmd.ftn
The GTYPE variable is originally used in calling the subroutine W3FLDO as an intended INOUT dummy argument, which may alter the GTYPE value if the forcing field did not set GTYPE correctly.   A copy of the GTYPE value is now substituted for the call so that GTYPE value remains as set in mod_def.ww3.

### Issue(s) addressed
Making SMC grid as an on-line option is a requested feature but there is no listed issue for it.
SMC multi-grid is a new feature.

### Check list  
* Is your feature branch up to date with the authoritative repository (NOAA/develop)?   Yes.
* Make sure you have checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop), [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop) and [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number)
* Reviewers: @mentions of suggested reviewers of the proposed changes.


### Testing
* How were these changes tested?
The branch is compiled on UKMO Cray XC40 machines with hybrid parallelization, with and without SMC switch.  Executables are tested with regular grid 50km global model and SMC36125 global models.  Multi-grid option with equal ranked SMC sub-grids, plus one regular sub-grid are also tested.

* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)
Three regression tests tp2.10, tp2.14, and tp2.16 are related to SMC grid and they may test the SMCTYPE and Arctic option.  
* If a new feature was added, was a new regression test added?
A new regression test will be added separately for the new SMC multi-grid option. 
* Have regression tests been run?  May need help for this work.
* Which compiler / HPC you used to run the regression tests in the PR? UKMO uses Cray XC40 machine.
* Please provide the summary output of matrix.comp (_matrix.Diff.out_, _matrixCompFull.out_ and _matrixCompSummary.out_):    
Please indicate the expected changes in the outputs ([excluding the known list of non-identical tests](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-master#4-look-at-results)).

ukmo-jianguo-li committed:
SMC multi-grid option changes.  JGLi11May2021

ukmo-jianguo-li committed:
SMC multi-grid changes for ww3_ounf.ftn   JGLi11May2021

ukmo-jianguo-li committed:
Add regression test mww3_test_09 for SMC multi-grid option.  JGLi12May2021

JessicaMeixner-NOAA commented on model/bin/make_makefile.sh:
```diff
@@ -327,11 +332,6 @@
                ID='rotated grid'
                TS='RTD'
                OK='RTD' ;;
-#sort:arctic:
```
the corresponding section in w3_new can also be removed for arctic

ukmo-jianguo-li replied:
RTD option is independent of SMC grid though they share some subroutines.

JessicaMeixner-NOAA replied:
The RTD option is still there with rotag, but the arctic and ARC switch are removed, correct?  If so, the "arctic" option could be removed from w3_new as well as being removed from make_makefile.sh

ukmo-jianguo-li replied:
Yes, the arctic option in w3_new could be removed as the ARC switch is removed.

ukmo-ccbunney replied:
Can you remove the "arctic" section from w3_new, please @ukmo-jianguo-li

aliabdolali replied:
thanks @ukmo-jianguo-li

JessicaMeixner-NOAA commented on model/ftn/w3wavemd.ftn:
```diff
@@ -1785,74 +1799,67 @@
+!/SMC           ENDIF 
+!/SMC 
+!/SMC  !Li    Work out root PE (ISPEC) and JSEA numbers for IY 
+!/SMC/!/DIST            ISPEC = MOD( IY-1, NAPROC ) 
```
Double switches is not really supported right now, so I'm not sure if this code segment will work as intended.

ukmo-ccbunney replied:
@JessicaMeixner-NOAA : PR #168 added multiple switch support last year, so this should work OK.

JessicaMeixner-NOAA replied:
Thanks for the reminder! That's great!

JessicaMeixner-NOAA commented on regtests/mww3_test_09/input/subGtLksgrid:
```diff
@@ -0,0 +1,87 @@
+#!/bin/bash
```
Should this stay in the main repo?  Also I couldn't comment on the .dat files themselves, but should these be in the repo or in the ftp bundle? GitHub wouldn't load many of the them so I wasn't sure what needed to be done there?

ukmo-jianguo-li replied:
These *.dat files are input files for the sample grid and they are better kept with the example programs.  They do not need any update so far.

ukmo-ccbunney replied:
I think this particular file should not be in the commit (it is a script to run the test on our HPC). I think you can remove this one @ukmo-jianguo-li .

As for the *.dat files, I would agree with @JessicaMeixner-NOAA - these should probably go in the ftp bundle. @ukmo-jianguo-li : when there are large input files associated with a regression test, the convention has been to put these in a separate "regtest data" tarball that is downloaded separately.

JessicaMeixner-NOAA commented on model/ftn/w3adatmd.ftn:
```diff
@@ -583,10 +584,10 @@
-!/SMC                                 NACT, NMXY
-!/SMC      INTEGER, POINTER        :: MAPX2(:), MAPY2(:), MAPAXY(:),       &
-!/SMC                                 MAPXY(:), MAPTH2(:), MAPWN2(:)
+!! !/SMC      INTEGER, POINTER        :: NMX0, NMX1, NMX2, NMY0, NMY1, NMY2,  &
```
Do you think we need to still keep these lines commented out, or could we go ahead and delete these?

ukmo-jianguo-li replied:
I forgot to tidy them up after finishing the SMC multi-grid changes.  Yes, they are safe to be deleted.

ukmo-jianguo-li commented:
All the conflicts are comment lines except for one.  They are easy to resolve but I do not know how to do it on the web page.  Could I leave them to any of you?  Thanks.

aliabdolali commented:
> All the conflicts are comment lines except for one. They are easy to resolve but I do not know how to do it on the web page. Could I leave them to any of you? Thanks.

You can delete them on your feature branch and it automatically will be reflected here.

ukmo-jianguo-li commented:
Yes, the ARC switch is removed and the arctic option in w3_new could be removed as well.

>The RTD option is still there with rotag, but the arctic and ARC switch are removed, correct? If so, the "arctic" option could be removed >from w3_new as well as being removed from make_makefile.sh

ukmo-jianguo-li commented:
The comment line conflicts are about date sequences.  They need to be altered in positions rather than be deleted.

aliabdolali commented:
@ukmo-jianguo-li Please send me the *.dat files and also update the model/bin/ww3_from_ftp.sh, so I will update the tar file and you can remove *.dat files from your feature branch.

ukmo-jianguo-li commented:
The regression test mww3_test_09 is not completed as I do not know how the regression tests work.  The 
subGtLksgrid job script is for reference to anyone who could help finish the setup.  It may be deleted if it is not needed.
I do not know how input data files are handled in the regression tests.  Anyone who helps complete the test setup, feel free to rearrange the files or put them in other places.
@aliabdolali Could you fetch the mww3_test_09/input/*.dat files from the branch?  What do you mean 'send the files to you'?  Do you like me email them to you?

ukmo-jianguo-li committed:
Merge branch 'develop' into fb_smc_multigrid

ukmo-jianguo-li committed:
Merge branch 'develop' into fb_smc_multigrid

ukmo-jianguo-li committed:
Delete *.dat files from regression test mw_09.

ukmo-ccbunney committed:
Update ww3_multi.inp to reflect recent changes to input field list.
Reduced spectral resolution of grids.

ukmo-ccbunney commented on model/ftn/wminitmd.ftn:
```diff
@@ -431,8 +432,12 @@
+      CHARACTER(LEN=13),                                              &
+                  ALLOCATABLE :: INAMES(:,:), MNAMES(:)
+      CHARACTER(LEN=40),                                              &
+                  ALLOCATABLE :: PNAMES(:)
```
I think something is not quite right here: INAMES, MNAMES and PNAMES are defined twice...

ukmo-ccbunney replied:
I suggest we just delete the lines you added @ukmo-jianguo-li . It looks like you just change the formatting so they are not split over multiple lines, I think...

ukmo-ccbunney commented on model/ftn/wminitmd.ftn:
```diff
@@ -3576,8 +3581,12 @@
+      CHARACTER(LEN=13)       :: STDATE, MN, TNAMES(9)
+      CHARACTER(LEN=40)       :: PN
+      CHARACTER(LEN=13),                                               &
+                  ALLOCATABLE :: INAMES(:,:), MNAMES(:)
```
The same issue here...

ukmo-ccbunney reviewed (changes requested):
A few issues with the conflict merge, I think...

ukmo-jianguo-li committed:
Remove repeated variable declaration lines.

ukmo-jianguo-li commented:
Duplicated lines deleted.  Changes committed and pushed into fb_smc_multigrid branch.

ukmo-ccbunney committed:
Updates to get mww3_test_09 running.

ukmo-ccbunney committed:
Converted grid.inp files to namelist versions

ukmo-jianguo-li committed:
Remove duplicated variable declarations (2).

ukmo-ccbunney committed:
Created namelist version of ww3_multi.
Cleaned up existing ww3_multi_grdset_a.inp file.

ukmo-ccbunney committed:
Renamed swithc file

ukmo-ccbunney commented:
@ukmo-jianguo-li I've updated the `mww3_test_09` regtest so that it runs in the regression test framework.
I have also converted the input files to namelist versions.

One thing to note: the model currently crashes when running using the namelist `ww3_multi.nml` due to an existing bug (see #378)

@aliabdolali - I can fix that bug in this PR if you like?

ukmo-ccbunney commented on model/ftn/wminitmd.ftn:
```diff
@@ -1460,7 +1463,7 @@
         IF(J .LE. 2) THEN
           OUTFF(J,I)=0
-          WORDS(1:6) =''
+          WORDS(6) =''
```
@ukmo-jianguo-li - did you mean to change this line?
This looks like an older version of the code that was changed in PR #166

ukmo-jianguo-li replied:
Do not remember I made that change.   Correct it please if you think it is wrong.

ukmo-ccbunney replied:
Done.

ukmo-ccbunney committed:
Reverted unintentional change.

ukmo-jianguo-li committed:
Tidy up commented lines.

ukmo-jianguo-li committed:
Merge branch 'fb_smc_multigrid' of https://github.com/ukmo-waves/WW3 into fb_smc_multigrid

ukmo-jianguo-li committed:
Merge branch 'develop' into fb_smc_multigrid
*** Warning: conflicts in model/ftn/ww3_ounf.ftn are not resolved yet!

ukmo-ccbunney committed:
Fixed conflicts in ww3_ounf. Renamed SMCTOUT variable to SMCOTYPE

ukmo-jianguo-li committed:
ww3_bound and ww3_bounc dependency and manual update.

aliabdolali commented:
@ukmo-jianguo-li 
I added the _*dat_ files to the tar file on the ftp. 
In your feature branch
https://github.com/ukmo-waves/WW3/blob/fb_smc_multigrid/model/bin/ww3_from_ftp.sh#L11
change the version number to _v7.12.3_
`ww3ver=v7.12.3`

and insert the following in 
https://github.com/ukmo-waves/WW3/blob/fb_smc_multigrid/model/bin/ww3_from_ftp.sh#L55
cp -r data_regtests/ww3_tp2.21/inputmww3_test_09/input/*     regtests/mww3_test_09/input/

ukmo-ccbunney committed:
Updated ww3_from_ftp to reference new tarfile

ukmo-ccbunney committed:
Merge branch 'develop' into fb_smc_multigrid
Fixed one merge conflict: model/ftn/w3gridmd.ftn

ukmo-ccbunney commented:
@ukmo-jianguo-li - I've updated the `ww3_from_ftp` file as per Ali's instructions above. I have also merged in the latest changes from `develop`, so this branch should be all up-to-date now.

ukmo-ccbunney commented on model/ftn/wmgridmd.ftn:
```diff
@@ -726,7 +747,10 @@
                DX_MAX_GRIDI=MAXVAL(GRIDS(I)%HPFAC)
                DY_MAX_GRIDI=MAXVAL(GRIDS(I)%HQFAC)
-            ELSEIF ( GRIDS(I)%GTYPE .EQ. RLGTYPE ) THEN
+!!Li        ELSEIF ( GRIDS(I)%GTYPE .EQ. RLGTYPE ) THEN
```
Delete commented out lines (just to keep the code tidy)

ukmo-ccbunney commented on model/ftn/wmgridmd.ftn:
```diff
@@ -774,7 +798,10 @@
                DX_MAX_GRIDJ=MAXVAL(GRIDS(J)%HPFAC)
                DY_MAX_GRIDJ=MAXVAL(GRIDS(J)%HQFAC)
-            ELSEIF ( GRIDS(J)%GTYPE .EQ. RLGTYPE ) THEN
+!!Li        ELSEIF ( GRIDS(J)%GTYPE .EQ. RLGTYPE ) THEN
```
Delete commented out lines (just to keep the code tidy)

ukmo-ccbunney commented on model/ftn/wmgridmd.ftn:
```diff
@@ -1307,7 +1334,10 @@
 
       DO GDST=1, NRGRD
-         IF ( GRIDS(GDST)%GTYPE .NE. RLGTYPE ) THEN
+!!Li     IF ( GRIDS(GDST)%GTYPE .NE. RLGTYPE ) THEN
```
Delete commented out lines (just to keep the code tidy)

ukmo-jianguo-li committed:
Update bin/w3_new to remove 'arctic' option.

ukmo-ccbunney commented on model/ftn/wmupdtmd.ftn:
```diff
@@ -637,16 +639,30 @@
 !
         CASE (2)
+!/SMC !!Li  For sea point current option FSWND.   JGLi08Feb2021
+!/SMC       IF( FSWND ) THEN
```
Is the "sea point wind" addition separate to the SMC mulitgrid additions?
If so, we might want to add this in it's own PR.

Have we got a regression test for sea point wind/current inputs?

EDIT: I see that `SEAWND` actually already exists in develop - so ignore that; I had not realised sea point only winds had already been added.

ukmo-jianguo-li committed:
Merge branch 'fb_smc_multigrid' of https://github.com/ukmo-waves/WW3 into fb_smc_multigrid

ukmo-ccbunney reviewed (changes requested):
@ukmo-jianguo-li - I've done a review of the code - there are a few comments from me.
Also - the new `mww3_test_09` regression tests needs adding to the regression test matrix (`regtests/bin/matrix.base`)

ukmo-ccbunney commented:
@ukmo-jianguo-li  - I am running a full regression test; I will update you with the results later.

ukmo-jianguo-li committed:
Tidy up commented lines in wmgridmd.ftn.

ukmo-ccbunney commented on model/bin/make_makefile.sh:
```diff
@@ -908,14 +905,14 @@
                core=
                data="w3adatmd $memcode w3gdatmd w3wdatmd w3idatmd w3odatmd"
-               prop=
+               prop="smcm"
```
should be `prop="$smcm"`

ukmo-ccbunney commented on model/bin/make_makefile.sh:
```diff
@@ -908,14 +905,14 @@
                core=
                data="w3adatmd $memcode w3gdatmd w3wdatmd w3idatmd w3odatmd"
-               prop=
+               prop="smcm"
```
Same here: should be `prop="$smcm"`

ukmo-ccbunney reviewed (changes requested):
A couple of bugs in make_makefile.sh

ukmo-jianguo-li committed:
Fix bug in make_makefile.sh

ukmo-ccbunney commented on model/ftn/wmgridmd.ftn:
```diff
@@ -4721,6 +4759,684 @@
+        CALL W3SETG( I, MDSE, MDST )
+!
+! Find local root PE and NAPROC for I grid.
+        ICROOT = MDATAS(I)%CROOT
```
@ukmo-jianguo-li - I am getting compile errors with `ww3_multi` when compiling as serial program (i.e. with the SHRD flag, rather than MPI DIST).

This subroutine (`WMSMCEQL`) accesses variable `CROOT`, but this variable is only defined when the MPI flag is set:

https://github.com/NOAA-EMC/WW3/blob/7db2f81f144bd882213798a62145c6cd44ce0442/model/ftn/wmmdatmd.ftn#L995

I think this subroutine needs some `!/MPI` switches added in appropriate places.

You can test this with one of the serial multi-grid regression tests, such as `mww3_test_02/PR1`

ukmo-ccbunney replied:
@ukmo-jianguo-li  There are other problems too with SHRD compilations:

This line is not needed and causes compilation errors in SHRD mode - I think it can simply be deleted:
https://github.com/NOAA-EMC/WW3/blob/7db2f81f144bd882213798a62145c6cd44ce0442/model/ftn/wmgridmd.ftn#L4857

This line is using CROOT too and needs to be behind an `MPI` switch:
https://github.com/NOAA-EMC/WW3/blob/7db2f81f144bd882213798a62145c6cd44ce0442/model/ftn/wminitmd.ftn#L2297

Can you fix these and run the `mww3_test_02` regtest (switch PR1) to check if there any on other issues? Looking quickly, I can see that there are likely more unprotected CROOT variables.

ukmo-ccbunney reviewed (changes requested):
@ukmo-jianguo-li  - I have found an issue with ww3_multi when compiled in SHRD mode.

ukmo-jianguo-li committed:
Shield CROOT with !/MPI switch.

ukmo-jianguo-li committed:
Remove bugs in compilation of ww3_multi with SHRD switch.

ukmo-jianguo-li committed:
Update tp2.10 switch files.

ukmo-jianguo-li committed:
Update switch files for tp2.16.

ukmo-jianguo-li committed:
Delete TNAMEP modification by SMC switch.

ukmo-juan-castillo committed:
Issue #398 (#28)

Added comments in code describing NRQMAX calculation
Fixes NOAA-EMC/WW3#398

ukmo-juan-castillo committed:
Changes to fix the bug reported in issue #390 (#26)

Update OASIS send coupling condition so that it works properly for longer runtimes.
Fixes NOAA-EMC/WW3#390

ukmo-ccbunney committed:
Updated missing PR2 and UNO switches in ww3_tp2.10 and ww3_tp2.14

ukmo-ccbunney committed:
Fixed some variable renames SMCTYPE => SMCOTYPE that were missed

ukmo-ccbunney committed:
Fixed missing GTYPE == SMCTYPE conditionals

ukmo-ccbunney commented:
Full regression test suite results.

There is just one oddity that I cannot explain (see details for ww3_t2.16 below). @ukmo-jianguo-li - can you comment?

The following changes are expected to be seen in SMC grid tests:
  - `mod_def.ww3` will have new `SMCTYPE` for GTYPE variable and extra SMC information.
  - `ww3_ounf.out` - a new string output "Conversion for SMCTYPE X" to stdout.
  - Extra `PR2` and `UNO` switch in the `WAVEWATCH_III_switches` global attribute of output netCDF files for some SMC grid tests
  - `ww3_grid.nml.log` - namelist options for ARCTIC now always written out (was previously only when !/ARC switch enabled). Also some extra namelist output due to explicit enabling of PR2 and UNO switches [**_will affect all tests run with -N flag_**]

Expected regression test differences due to above:
  - **ww3_ta1** 
  - **ww3_tnc1**
  - **ww3_tp2.10/MPI_OMP**:
   - **ww3_tp2.14/OASACM6**  
   - **ww3_tp2.16/** @ukmo-jianguo-li : The `WND`  values (all zero for this run) have a different SIGN in the ww3.*.wnd output files, i.e. some are +0 and some are -0. Although +0 or -0 are essentially the same, I don't know why this difference is occurring.
   - **ww3_tr1**
   - **ww3_ts1**



Usual known non B4B tests:
   - mww3_test_03/
   - ww3_tp2.10/./work_MPI_OMPH  
   - ww3_tp2.18/./work_TIDE_MPI

ukmo-jianguo-li commented:
> There is just one oddity that I cannot explain (see details for ww3_t2.16 below). @ukmo-jianguo-li - can you comment?
> 
> The following changes are expected to be seen in SMC grid tests:
>
>     * `ww3_grid.nml.log` - namelist options for ARCTIC now always written out (was previously only when !/ARC switch enabled). 

The ARC switch is merged into SMC switch and a namelist variable ARCTIC is used to control whether the Arctic part is included.  As ARCTIC is set to be .TRUE. in the ww3_grid.inp now the Arctic part is selected.

ukmo-ccbunney commented:
> 
> 
> > There is just one oddity that I cannot explain (see details for ww3_t2.16 below). @ukmo-jianguo-li - can you comment?
> 
> The ARC switch is merged into SMC switch and a namelist variable ARCTIC is used to control whether the Arctic part is included. As ARCTIC is set to be .TRUE. in the ww3_grid.inp now the Arctic part is selected.

The oddity I mean is the difference in the sign of zero values ww3*.wnd files.

ukmo-jianguo-li commented:
I do not have any idea about the +0 and -0 changes in tp2.16 test.  As far as I remember, there is no change to the regular grid wind for this update.

ukmo-ccbunney committed:
Merge branch 'NOAA-EMC:develop' into develop

ukmo-ccbunney committed:
Merge remote-tracking branch 'origin/develop' into fb_smc_multigrid

ukmo-jianguo-li committed:
Delete tp2.16 test for ARC swith as ARC is merged into SMC switch.

ukmo-jianguo-li committed:
Merge branch 'fb_smc_multigrid' of https://github.com/ukmo-waves/WW3 into fb_smc_multigrid

ukmo-jianguo-li commented:
@aliabdolali @ukmo-ccbunney  As the ARC switch is merged into the SMC one, the test tp2.16 for the ARC switch is no longer needed.  I have deleted the input files but do not know how to modify the job scripts for the regression tests.  Could any of you help tidy up the regression job scripts to complete the tp2.16 removal?
The +0 and -0 change in the tp2.16 test does not affect the model performance, I think we can go ahead to finish the merge first.  If Chris is still interested to find out what caused this change, it may be investigated later in the new branch.

ukmo-ccbunney commented:
> 
> 
> @aliabdolali @ukmo-ccbunney As the ARC switch is merged into the SMC one, the test tp2.16 for the ARC switch is no longer needed. I have deleted the input files but do not know how to modify the job scripts for the regression tests. Could any of you help tidy up the regression job scripts to complete the tp2.16 removal?
> The +0 and -0 change in the tp2.16 test does not affect the model performance, I think we can go ahead to finish the merge first. If Chris is still interested to find out what caused this change, it may be investigated later in the new branch.

Hi @ukmo-jianguo-li : Although the ARC compilation switch has been removed, it is still a feature of the SMC grid. Do you not think we should keep the ww3_tp2.16 regression test as it is the only regtest that runs with ARCTIC capability of the SMC grid enabled?

ukmo-jianguo-li commented:
@ukmo-ccbunney There is no need to keep the tp2.16 test.  The Arctic part is now tested along with other PSMC namelist options, like SWND and AVERG, in our local tests when SMC switch is selected.

JessicaMeixner-NOAA commented:
> @ukmo-ccbunney There is no need to keep the tp2.16 test. The Arctic part is now tested along with other PSMC namelist options, like SWND and AVERG, in our local tests when SMC switch is selected.

Which local test?  We need a test in the regression tests that test the arctic feature so that when new updates to WW3 come in, we can ensure that they do not break this feature.

ukmo-jianguo-li commented:
@JessicaMeixner-NOAA The Arctic part is now part of the SMC switch option.  As long as SMC switch is tested, it will be fine.  All features within the SMC switch are tested with our own tests, including other terms.

JessicaMeixner-NOAA commented:
@ukmo-jianguo-li to make sure I understand, there is another regression test in which the arctic region/domain is also included in another test?   

If no, then I think I agree with @ukmo-ccbunney that we need to keep the test. 
If yes, then we can just remove the whole folder, but then we need to update all of the manual and other places as well if we are removing the test completely .

ukmo-jianguo-li commented:
@JessicaMeixner-NOAA The regression test tp2.16 only checks the ARC switch works.  It does not test the Arctic part works correctly.  We have to use non-zero wind forcing and source terms together to test the Arctic part.  It does not matter whether the Arctic part is included or not in SMC switch test, like tp2.14.

ukmo-ccbunney commented:
@ukmo-jianguo-li The ww3_tp2.16 test, as you say, does not have any source terms, but it does test propagation in the arctic using a restart file generated with and idealised Gaussian spectrum. I think it is important we keep this regression test as it does test the arctic parts of the SMC code when the runtime ARCTIC option is set (as it is in this test). I know that the ARC switch no longer exists, but the ARCTIC code still exists and still needs to be tested.

I might also be a good idea to enable the source terms and provide a homogenous wind field in `ww3_shel.inp/nml` - that would be quite easy to do and would provide a more complete test (propagation + source terms).

What do you think?

ukmo-jianguo-li commented:
@ukmo-ccbunney The tp2.16 test was added to test the ARC switch when the Arctic part is included.  As the ARC switch is now merged into the SMC switch, there is no need for this test as one SMC switch test (tp2.14) will be enough to test the old ARC and SMC switches together.  If you like to include wind and source terms for test of the Arctic part physics, it will be a new test and you can do it in a new branch.  This branch is aimed for adding the multi-grid feature, not for testing the Arctic part.

JessicaMeixner-NOAA commented:
I agree that keeping the tp2.16 test (and possibly adding to it in a new PR later) sounds like a good idea.

aliabdolali commented:
The two tests for **ww3_tp2.16** compile and run times are:
 **MPI:**
_Total compile time:   437.05 sec
 Total run time    :    45.41 sec_
and 
**MPI_OMPH**
 _Total compile time:   862.99 sec
 Total run time    :    88.39 sec_
so it does not add much to our regtesting time. I think we can keep it as it.

ukmo-ccbunney commented:
> _Total compile time: 437.05 sec 
> _Total compile time: 862.99 sec 

😲 Are those compile times really that high?!

aliabdolali commented:
> > _Total compile time: 437.05 sec
> > _Total compile time: 862.99 sec
> 
> 😲 Are those compile times really that high?!

How it looks like on your HPC?
I remember Jessica prepared a spreadsheet with compile and run times for all the tests.

ukmo-ccbunney commented:
> 
> 
> > > _Total compile time: 437.05 sec
> > > _Total compile time: 862.99 sec
> > 
> > 
> > 😲 Are those compile times really that high?!
> 
> How it looks like on your HPC?
> I remember Jessica prepared a spreadsheet with compile and run times for all the tests.

For ww3_tp2.16 building ww3_grid, ww3_strt, ww3_ounf, ww3_ounp and ww3_shel (MPI) from clean takes a total of 125s on our GNU compiler 

Saying that... if I compile with our Cray compiler it is significantly slower, so maybe your numbers are not so surprising after all...

ukmo-ccbunney commented:
So, I've been looking into the issue of the differences between the **wind output** for `ww3_tp2.16` and have got a bit further (although I can't explain it yet).

I have tried setting a non-zero homogenous wind for the test so I could check if  it was something to do with having zero winds.
The results show that the output wind fields produced from this `fb_smc_multigrid` branch (top panel) look sensible, but the wind fields from the `develop` branch (bottom panel) look **wrong**.  _Note - the outputs will not look like homogenous as the arctic part of the SMC grid has a rotation applied._

![image](https://user-images.githubusercontent.com/48915820/123108589-73bab000-d432-11eb-8706-ed2a5dfd94b9.png)
_V-wind: fb_smc_multigrid (top), develop (bottom)_

I then enabled the **source terms** to see if this was a problem with the input winds, but the significant wave heights between the two tests are identical, so the problem lies just in the output of the wind field.

I then tried forcing with homogenous currents to see if they behaved the same way, but both output current fields look the same (and look correct).

![image](https://user-images.githubusercontent.com/48915820/123108831-a9f82f80-d432-11eb-93ba-de1360029cbc.png)
_V-current:fb_smc_multigrid (top), develop (bottom)_

I've not yet got to the bottom of the reason for the wind output differences - I will keep digging, but any suggestions at this point will be welcome! @ukmo-jianguo-li - could this be related to "sea point only" winds?

_**However, the good news is that the fb_smc_multigrid branch appears to be fixing a pre-existing issue with wind output on the arctic part of the SMC grid**_

ukmo-ccbunney commented:
OK - I know where the differences are coming from now.

There are some `!/ARC` switches still in the code. Specifically, some in `w3iogomd`:
https://github.com/ukmo-waves/WW3/blob/0fa151e3dbdc4ff152cda8c5b51525ec96f8eb40/model/ftn/w3iogomd.ftn#L2844-L2848
Cut and pasted below:
```fortran
                    DO ISEA=1, NSEA
!/ARC  !!Li  Rotate map-east wind in Arctic part back to local east.  JGLi02Feb2016
!/ARC                      IF(ISEA .GT. NGLO) THEN
!/ARC                        UDARC = UD(ISEA) - ANGARC(ISEA - NGLO)*DERA
!/ARC                        UD(ISEA) = MOD(TPI + UDARC, TPI)
!/ARC                      ENDIF
```

If I remove these switches and wrap the code block in a `IF(ARCTC)` conditional:
```fortran
                  DO ISEA=1, NSEA
                      IF(ARCTC) THEN
                        !!Li  Rotate map-east wind in Arctic part back to local east.  JGLi02Feb2016
                        IF(ISEA .GT. NGLO) THEN
                          UDARC = UD(ISEA) - ANGARC(ISEA - NGLO)*DERA
                          UD(ISEA) = MOD(TPI + UDARC, TPI)
                        ENDIF
                      ENDIF
```

then I get the same results as the `develop` branch.

**However, these results look wrong to me!**

 So - @ukmo-jianguo-li : can you have a look and see whether those rotation lines need to be there still, as they actually seem to corrupt the wind output.

Also note that there are some other places that still have `!/ARC` switches:
   - w3iogomd: https://github.com/ukmo-waves/WW3/blob/0fa151e3dbdc4ff152cda8c5b51525ec96f8eb40/model/ftn/w3iogomd.ftn#L2462
   - w3iogomd: https://github.com/ukmo-waves/WW3/blob/0fa151e3dbdc4ff152cda8c5b51525ec96f8eb40/model/ftn/w3iogomd.ftn#L2869-L2873
   - w3updmd: https://github.com/ukmo-waves/WW3/blob/0fa151e3dbdc4ff152cda8c5b51525ec96f8eb40/model/ftn/w3updtmd.ftn#L733
   - w3updmd: https://github.com/ukmo-waves/WW3/blob/0fa151e3dbdc4ff152cda8c5b51525ec96f8eb40/model/ftn/w3iogomd.ftn#L2844-L2848

ukmo-jianguo-li commented:
@ukmo-ccbunney Great, you can dig into the problem in the regression test.  As far as I remember, there is no intended change for the wind output in this update.  If you could find any, I'll be interested to know.  The sea-point wind option only affects input wind.  After the regular wind is interpolated to sea-points inside the model, the two options will be the same.  Sea-point input wind does not need any spatial interpolation inside the model as it is already on sea-points.  Output wind is based on the interpolated sea-point wind so the two options should have no difference.  All my tests only check Hs output so I do not know any changes in output wind.

JessicaMeixner-NOAA commented:
This could be completely unrelated, but we've had issues before where if we start from a restart run the wind output for the first time step is different even though all the computed parameters are the same.  @aliabdolali @ajhenrique myself and others have looked into this with no success.  I don't think this is necessarily related, but thought I'd put it out there just in case.  We looked a lot at how the cos() of values were being computed but nothing fixed the issue.  I think it's different but some of @ukmo-ccbunney  description of the debugging made me think it could be similar/same?

ukmo-ccbunney commented:
> 
> 
> This could be completely unrelated, but we've had issues before where if we start from a restart run the wind output for the first time step is different even though all the computed parameters are the same. @aliabdolali @ajhenrique myself and others have looked into this with no success. I don't think this is necessarily related, but thought I'd put it out there just in case. We looked a lot at how the cos() of values were being computed but nothing fixed the issue. I think it's different but some of @ukmo-ccbunney description of the debugging made me think it could be similar/same?

Thanks @JessicaMeixner-NOAA . I had the same issue when I started from cold too (i.e. no restart file). Anyhow - I think I have got to the bottom of this now. See my comment above.

JessicaMeixner-NOAA commented:
I missed that comment, thanks! Glad it's resolved

ukmo-jianguo-li commented:
@ukmo-ccbunney I just looked at the two modules and there are some old ARC lines which crept back with a recent merge for a new output variable TAU.  Because they only affect output variables they slipped through my tests.  I have updated the two modules and will try to re-do my tests.  I do not think they will affect the Hs output.  As the code compiles fine I'll just commit the changes so you can go on with your regression test.  This is a typical problem when other merge intervenes before the branch is merged.
The lines to rotate the wind vectors back to the local east system for the Arctic part should be correct.  If you suspect the output is wrong, I may have a close look later and fix it in a new bug-fixing branch if necessary.  Now I simply keep them for this merge.

ukmo-jianguo-li committed:
Rectify old ARC lines brought back with a recent merge.

ukmo-ccbunney committed:
Revert "Delete tp2.16 test for ARC swith as ARC is merged into SMC switch."

This reverts commit 2da27d3917adc24983f4752396115c9421c58d1b.

ukmo-ccbunney committed:
Merge remote-tracking branch 'upstream/develop' into fb_smc_multigrid

ukmo-ccbunney commented:
I've reverted the commit that removed the ww3_tp2.16 regression tests - these are all consistent with the develop branch now. I will raise a separate issue to address the potential bug in the arctic wind output.

I believe this PR is now ready for a final review.

I will re-run the regression tests and post the results here ASAP.

ukmo-ccbunney commented:
Results from regression test:
[matrixComp_fb_smc_multigrid_20210625.zip](https://github.com/NOAA-EMC/WW3/files/6717036/matrixComp_fb_smc_multigrid_20210625.zip)


**Non-identical cases:**

**Existing:**
```
mww3_test_03/./work_PR3_UQ_MPI_d2                     (10 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (13 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (9 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (10 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (9 files differ)
ww3_tp2.10/./work_MPI                     (4 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (4 files differ)
```

**Expected differences from this PR:**

- ww3_ta1, ww3_tnc1, ww3_ts1/./work_ST4_T700:
  -  ww3_grid.mnl.log
- ww3_tp2.14/./work_OASACM6, ww3_tp2.16:
  - ww3_grid.out: slight difference in some of the output info
  - ww3_ounf.out: New "Conversion for SMCTYPE:" message
  - ww3.*.nc: removal of "ARC" switch in `WWATCH_III_Switches` attribute


NOTE: The following two test were showing differences, but they disappeared when I re-ran the tests:
- ww3_tp2.5/./work_REF_PR3_UQ_MPI
   - Differences in out_grd and some *.fp and *.spr files
- ww3_tr1:
  - Differences in mod_def.ww3

This happens to me occasionally with the regression tests and I don't have an explanation for it! Perhaps some modules are not being correctly touched when the switch file changes?

I suspect for ww3_tr1, there is an uninitialised variable somewhere, I am investigating. Edit: Issue #425 opened for this.

ukmo-ccbunney committed:
Merge branch 'develop' into fb_smc_multigrid

aliabdolali reviewed (commented):
Hi @ukmo-jianguo-li 
Here are a couple of comments I have on your PR:
1- manual needs updates: to include SMC type=1,2,3,4 and cleaning ARC switch. 
2- the new regtest mww3_test09 is not added to matrix_ncep, matrix_ukmo, ...

ukmo-jianguo-li committed:
Merge branch 'develop' into fb_smc_multigrid

ukmo-jianguo-li commented:
> Hi @ukmo-jianguo-li
> Here are a couple of comments I have on your PR:
> 1- manual needs updates: to include SMC type=1,2,3,4 and cleaning ARC switch.
> 2- the new regtest mww3_test09 is not added to matrix_ncep, matrix_ukmo, ...
@aliabdolali Hi, Ali. The manual page for the SMC grid part has been updated.  Which part you mean for adding the SMC type and cleaning ARC switch? 
The regression test is handled by Chris, who is on leave now.  I do not know how to add it to your matrix file.
The conflict in w3gdatmd.ftn is about a newly added log line.  I tried to solve it online and commit the change but may not be successful.

aliabdolali commented:
Here is the summary of regtests on NOAA hpc.
Note that the mww3_test09 is not in the matrix, so I added a comment in my review.

The summary:
**The following differences are expected from non-b4b list.**
```
mww3_test_03/./work_PR2_UQ_MPI_d2                     (9 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (9 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (9 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (7 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (8 files differ)
mww3_test_07/./work_PR3_UQ                     (3 files differ)
ww3_tp2.15/./work_5km                     (1 files differ)
ww3_tp2.15/./work_MPI_5km                     (1 files differ)

```
**the following are expected due to changes in the switch and definitions.**  could you confirm @ukmo-jianguo-li  and @ukmo-ccbunney 
```
ww3_tp2.10/./work_MPI_OMPH                     (9 files differ)
ww3_tp2.10/./work_MPI                     (4 files differ)
ww3_tp2.14/./work_OASACM6                     (4 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (5 files differ)
ww3_tp2.16/./work_MPI                     (5 files differ)
ww3_ts1/./work_ST4_T700                     (1 files differ)
ww3_ta1/./work_UPD3_U_cap                     (1 files differ)
ww3_ta1/./work_UPD0F_O                     (1 files differ)
ww3_ta1/./work_UPD6_O                     (1 files differ)
ww3_ta1/./work_UPD5_U                     (1 files differ)
ww3_ta1/./work_UPD2_U                     (1 files differ)
ww3_tnc1/./work                     (1 files differ)

```
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6849524/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6849525/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/6849526/matrixDiff.txt)

ukmo-jianguo-li commented:
@aliabdolali I do not know much about the regression tests so please wait for @ukmo-ccbunney  for the confirmation.

aliabdolali commented:
> > Hi @ukmo-jianguo-li
> > Here are a couple of comments I have on your PR:
> > 1- manual needs updates: to include SMC type=1,2,3,4 and cleaning ARC switch.
> > 2- the new regtest mww3_test09 is not added to matrix_ncep, matrix_ukmo, ...
> > @aliabdolali Hi, Ali. The manual page for the SMC grid part has been updated.  Which part you mean for adding the SMC type and cleaning ARC switch?
> > The regression test is handled by Chris, who is on leave now.  I do not know how to add it to your matrix file.
> > The conflict in w3gdatmd.ftn is about a newly added log line.  I tried to solve it online and commit the change but may not be successful.

Please modify line 313 of **manual/impl/switch.tex** as ARC switch is removed.
Please add _mww3_test_09_ to **manual/impl/test_it.tex**

ukmo-ccbunney committed:
Added mww3_test_09 to the regression test matrix

ukmo-ccbunney commented:
@aliabdolali 
>Note that the mww3_test09 is not in the matrix, so I added a comment in my review.

I've updated this in commit 951b779

ukmo-ccbunney commented:
> the following are expected due to changes in the switch and definitions. could you confirm @ukmo-jianguo-li and @ukmo-ccbunney
ww3_tp2.10/./work_MPI_OMPH                     (9 files differ)
ww3_tp2.10/./work_MPI                     (4 files differ)
ww3_tp2.14/./work_OASACM6                     (4 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (5 files differ)
ww3_tp2.16/./work_MPI                     (5 files differ)
ww3_ts1/./work_ST4_T700                     (1 files differ)
ww3_ta1/./work_UPD3_U_cap                     (1 files differ)
ww3_ta1/./work_UPD0F_O                     (1 files differ)
ww3_ta1/./work_UPD6_O                     (1 files differ)
ww3_ta1/./work_UPD5_U                     (1 files differ)
ww3_ta1/./work_UPD2_U                     (1 files differ)
ww3_tnc1/./work                     (1 files differ)

@aliabdolali - I've look at matrixDiff.txt and those changes look fine; they are all as expected and as described in comment above https://github.com/NOAA-EMC/WW3/pull/372#issuecomment-861540736 (disregarding the fixed issue with ww3_tp2.16 WND values).

ukmo-jianguo-li committed:
Modify manual/impl/ files for SMC grid update.

ukmo-jianguo-li committed:
Merge branch 'fb_smc_multigrid' of https://github.com/ukmo-waves/WW3 into fb_smc_multigrid

ukmo-jianguo-li commented:
@aliabdolali Done the manual changes and please see whether they are correct.  I'm not quite confident with the Latex files.

aliabdolali commented:
I am going to check this PR. @ukmo-ccbunney could you merge develop and fix the conflict?

aliabdolali commented on model/bin/ww3_from_ftp.sh:
```diff
@@ -53,6 +53,7 @@ cp -r data_regtests/ww3_tp2.14/input/toy/toy_coupled_field.nc.OASACM regtests/ww
 cp -r data_regtests/ww3_tp2.14/input/toy/*.nc regtests/ww3_tp2.14/input/toy/
 cp -r data_regtests/ww3_tp2.17/input/*     regtests/ww3_tp2.17/input/
 cp -r data_regtests/ww3_tp2.21/input/*     regtests/ww3_tp2.21/input/
+cp -r data_regtests/ww3_test_09/input/*    regtests/mww3_test_09/input/
```
please correct it:
cp -r data_regtests/mww3_test_09/input/* regtests/mww3_test_09/input/

aliabdolali commented on regtests/bin/matrix.base:
```diff
@@ -1903,6 +1904,12 @@
+  if [ "$multi09" = 'y' ]
+  then
+    echo ' ' >> matrix.body
+    echo "$rtst -s MPI -w work_MPI -m grdset_a -f -p $mpi -n $np $ww3 mww3_test_09" >> matrix.body
```
"-N" is missing for namelist
echo "$rtst -s MPI -w work_MPI -m grdset_a -f -p $mpi -n $np -N $ww3 mww3_test_09" >> matrix.body

ukmo-ccbunney replied:
@aliabdolali I thought it was normal to not specify the -N switch so that the regression test user can set it via the $rtst variable. I.e. rtst="run_test -N".

I am happy to specify it explicitly in the matrix.base if you prefer though (I guess we are deprecating the .inp files...)

aliabdolali replied:
You are right, however, I was not able to run it with ww3_grid.<>.nml files. The run_test did not recognize them and it failed. Have you tested them at your end without -N?

ukmo-ccbunney replied:
So - the NML option (-N) works for me and the test runs ok (once I had installed the data using ww3_from_ftp).
I thought I had set up .inp files for the ww3_grid program, but perhaps not. Would you like me to create these, considering that we are deprecating .inp files?

aliabdolali replied:
> So - the NML option (-N) works for me and the test runs ok (once I had installed the data using ww3_from_ftp).
> I thought I had set up .inp files for the ww3_grid program, but perhaps not. Would you like me to create these, considering that we are deprecating .inp files?

So we either need to specifiy -N in matrix or add grid<>.inp files there. Since we do not have -N in the matrix for none of other tests, I would prefer to have *inp files for this one as well.

ukmo-ccbunney replied:
OK - I will prepare some .inp files.

ukmo-ccbunney replied:
@aliabdolali ww3_grid<>.inp files have been added for mww3_test_09.
I've tested running with both the .nml and .inp files and both are running ok for me.

ukmo-ccbunney committed:
Merge remote-tracking branch 'origin/develop' into fb_smc_multigrid

ukmo-ccbunney committed:
Fixed typo in ww3_from_ftp.sh

ukmo-ccbunney committed:
Added ww3_grid_<..>.inp files for mww3_test_09

aliabdolali commented:
@ukmo-ccbunney thanks, I also checked it and the test ran successfully.

ukmo-ccbunney commented:
Thanks for all the hard work on getting this one merged @aliabdolali 👍
